## Gather

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [3]:
patients.head(3)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   

                                 contact  birthdate  weight  height   bmi  
0   951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1  PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2       402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8

In [1]:
treatments.head(3)

NameError: name 'treatments' is not defined

In [36]:
adverse_reactions.head(3)

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia

In [37]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [11]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [12]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [38]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [17]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [39]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [20]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [21]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [22]:
patients.sample(5)

patient_id assigned_sex given_name   surname                 address  \
58           59       female     Yasmin    Araujo      3682 Stiles Street   
176         177       female      Makda   Abraham       4093 Smith Street   
395         396         male  Krisztián     Bakos        2316 Ingram Road   
280         281         male      David  Beauvais      98 Freshour Circle   
468         469         male       Tosh    Jensen  1536 John Daniel Drive   

            city state  zip_code        country  \
58    Pittsburgh    PA   15205.0  United States   
176       Lowell    MA    1852.0  United States   
395  Yadkinville    NC   27055.0  United States   
280  San Antonio    TX   78258.0  United States   
468       Argyle    MO   65001.0  United States   

                                           contact   birthdate  weight  \
58   YasminCardosoAraujo@superrito.com412-640-7035   3/21/1981   196.0   
176          MakdaAbraham@einrot.com1 508 857 0477   6/24/1979   145.2   
395         336-677-8769BakosKrisztian@teleworm.us   3/28/1975   231.9   
280           210-218-3477DavidLBeauvais@rhyta.com    5/1/1931   128.5   
468          573-728-3861ToshCJensen@superrito.com  12/19/1969   194.7   

     height   bmi  
58       67  30.7  
176      65  24.2  
395      74  29.8  
280      67  20.1  
468      68  29.6

In [23]:
patients.surname.value_counts()

Doe              6
Jakobsen         3
Taylor           3
Tucker           2
Dratchev         2
Aranda           2
Kadyrov          2
Lund             2
Parker           2
Schiavone        2
Tạ               2
Gersten          2
Woźniak          2
Lương            2
Johnson          2
Cabrera          2
Batukayev        2
Kowalczyk        2
Lâm              2
Silva            2
Nilsen           2
Berg             2
Cindrić          2
Ogochukwu        2
Bùi              2
Grímsdóttir      2
Liễu             2
Hueber           2
Correia          2
Souza            2
                ..
Sjöberg          1
Mehler           1
Glaser           1
Lei              1
Compagnon        1
Zaitseva         1
Johnsrud         1
Lê               1
Henriksen        1
Teichelmann      1
Novosel          1
Zhdanov          1
Györfy           1
Nash             1
Lange            1
Uspenskaya       1
Saenger          1
Nakahara         1
Wong             1
Villadsen        1
Wiśniewska       1
Maslov      

In [24]:
patients.address.value_counts()

123 Main Street              6
648 Old Dear Lane            2
2778 North Avenue            2
2476 Fulton Street           2
2931 Romano Street           1
1051 Bassell Avenue          1
550 Cliffside Drive          1
576 Brown Bear Drive         1
602 Tator Patch Road         1
2935 Diamond Cove            1
901 Southern Street          1
1840 Millbrook Road          1
1049 Park Boulevard          1
4102 Ritter Avenue           1
1067 Ray Court               1
4988 Lynn Street             1
577 Chipmunk Lane            1
3300 Woodridge Lane          1
2055 Emeral Dreams Drive     1
3730 Seth Street             1
4992 Moonlight Drive         1
284 Mudlick Road             1
4120 Thompson Drive          1
1829 Bungalow Road           1
4064 Marigold Lane           1
2640 Sweetwood Drive         1
3479 Elm Drive               1
3130 Jessie Street           1
4110 Capitol Avenue          1
2172 Lynn Street             1
                            ..
1257 Elsie Drive             1
4853 Uni

In [25]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [26]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [27]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [28]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [29]:
sum(treatments.auralin.isnull())

0

In [30]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [4]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

<font color='red'>Complete the following two "Missing Data" **Define, Code, and Test** sequences after watching the *"Address Missing Data First"* video.</font>

#### `treatments`: Missing records (280 instead of 350)

##### Define
* We can't get this data as of now. So we will append the patients in this dataset and leave it blank. 
* Note: the missing `treatments` records are stored in a file named `treatments_cut.csv`, which you can see in this Jupyter Notebook's dashboard (click the **jupyter** logo in the top lefthand corner of this Notebook). Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) for the function used in the solution.*

##### Code

In [40]:
treatments_cut = pd.read_csv('treatments_cut.csv')

In [55]:
treatments_clean = pd.concat([treatments , treatments_cut] ,ignore_index=True)

##### Test

In [19]:
treatments_clean

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
320      eufemio       rosario          -  37u - 40u         7.54       7.26   
321     dalmacia        madrid          -  26u - 23u         7.67       7.21   
322        daimy         tromp          -  40u - 45u         9.41       8.94   
323   hans peter      radislav  42u - 53u          -         7.60       7.20   
324       jeremy       montagu          -  52u - 52u         7.68       7.36   
325    štefanija       novosel  37u - 42u          -         7.64       7.17   
326         anja       mueller  43u - 56u          -         7.58       7.29   
327       regolo         nucci  51u - 59u          -         7.53       7.02   
328      nebechi     ekechukwu          -  37u - 39u         7.78       7.39   
329      satsita     batukayev          -  42u - 42u         7.63       7.25   
330      timothy        cotton          -  26u - 25u         7.92       7.52   
331      bjørnar        nilsen          -  36u - 33u         7.99       7.70   
332       vivien        zsinkó  30u - 38u          -         8.35       7.92   
333        genet      fesahaye  25u - 35u          -         7.73       7.36   
334        borna     šlezinger          -  42u - 41u         7.55       7.18   
335         mary         adams          -  32u - 33u         7.65       7.26   
336       blanka      jurković  31u - 43u          -         7.77       7.33   
337       joseph        tucker  48u - 56u          -         7.67       7.30   
338  c

In [66]:
treatments_clean.shape

(350, 7)

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

##### Define
* we will recalculate HbA1c change from HbA1c_start HbA1c_end

##### Code

In [56]:
treatments_clean['hba1c_change'] = treatments_clean['hba1c_start'] - treatments_clean['hba1c_end']

##### Test

In [68]:
treatments_clean

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
320      eufemio       rosario          -  37u - 40u         7.54       7.26   
321     dalmacia        madrid          -  26u - 23u         7.67       7.21   
322        daimy         tromp          -  40u - 45u         9.41       8.94   
323   hans peter      radislav  42u - 53u          -         7.60       7.20   
324       jeremy       montagu          -  52u - 52u         7.68       7.36   
325    štefanija       novosel  37u - 42u          -         7.64       7.17   
326         anja       mueller  43u - 56u          -         7.58       7.29   
327       regolo         nucci  51u - 59u          -         7.53       7.02   
328      nebechi     ekechukwu          -  37u - 39u         7.78       7.39   
329      satsita     batukayev          -  42u - 42u         7.63       7.25   
330      timothy        cotton          -  26u - 25u         7.92       7.52   
331      bjørnar        nilsen          -  36u - 33u         7.99       7.70   
332       vivien        zsinkó  30u - 38u          -         8.35       7.92   
333        genet      fesahaye  25u - 35u          -         7.73       7.36   
334        borna     šlezinger          -  42u - 41u         7.55       7.18   
335         mary         adams          -  32u - 33u         7.65       7.26   
336       blanka      jurković  31u - 43u          -         7.77       7.33   
337       joseph        tucker  48u - 56u          -         7.67       7.30   
338  c

### Tidiness

<font color='red'>Complete the following four "Tidiness" **Define, Code, and Test** sequences after watching the *"Cleaning for Tidiness"* video.</font>

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
* Extract the phone number and email from contact column and split into 2 columns

*Your definition here. Hint 1: use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). Hint 2: [various phone number regex patterns](https://stackoverflow.com/questions/16699007/regular-expression-to-match-standard-10-digit-phone-number). Hint 3: [email address regex pattern](http://emailregex.com/), which you might need to modify to distinguish the email from the phone number.*

##### Code

In [21]:
patients['contact']

0                   951-719-9170ZoeWellish@superrito.com
1                  PamelaSHill@cuvox.de+1 (217) 569-3204
2                       402-363-6804JaeMDebord@gustr.com
3             PhanBaLiem@jourrapide.com+1 (732) 636-8246
4                        334-515-7487TimNeudorf@cuvox.de
5               386-334-5237RafaelCardosoCosta@gustr.com
6                      775-533-5933MaryBAdams@einrot.com
7                   XiuxiuChang@einrot.com1 408 778 3236
8                816-265-9578DavidGustafsson@armyspy.com
9          SophieCabreraIbarra@teleworm.us1 718 795 9124
10                906-478-8949SandyGunnarsson@dayrep.com
11            Abdul-NurMummarIsa@rhyta.com1 931 207 0839
12             OmeokachieIbeamaka@einrot.com434-509-2614
13            AnenechiChidi@armyspy.com+1 (205) 417-8095
14                     AsiaWozniak@rhyta.com918-712-3469
15                       276-225-1955SrenFLund@gustr.com
16                LieuThiThuTam@dayrep.com1 559 765 7836
17              RoxanneAndreyev

In [9]:
patients_clean['phone'] = patients_clean['contact'].str.extract(r"((\+?\d{1,2}\s)?\(?\d{3}\)?[\s.-]\d{3}[\s.-]\d{4})", expand = True)[0]

In [ ]:
patients_clean

In [10]:
def remove_ph(row):
    '''
    replace phone from contact field. left over will be email
    '''
    ph = str(row['phone'])
    email = str(row['contact']).replace(ph, '')
    return email

In [11]:
patients_clean['email'] = patients_clean.apply(remove_ph, axis=1)

##### Test

In [62]:
patients_clean.head(5)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height  \
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66   
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66   
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71   
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70   
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27   

    bmi              phone                      email  
0  19.6       951-719-9170   ZoeWellish@superrito.com  
1  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2  24.8       402-363-6804       JaeMDebord@gustr.com  
3  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4  26.1       334-515-7487        TimNeudorf@cuvox.de

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
* Create a column that will contain which treatment was given - Auralin, Novodora
* Create 2 columns that will contain start doese and end dose for the treatment

*Your definition here. Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html). Here is an excellent [`melt` tutorial](https://deparkes.co.uk/2016/10/28/reshape-pandas-data-with-melt/).*

##### Code

In [13]:
treatments_clean.head(3)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43

In [57]:
treatments_clean= treatments_clean.melt(id_vars = ['given_name', 'surname','hba1c_start', 'hba1c_end','hba1c_change'], 
                                         var_name = 'treatment', value_name='dose')
treatments_clean = treatments_clean[treatments_clean['dose'] !='-']
treatments_clean['dose_start'] = treatments_clean['dose'].str.split(pat = '-', expand = True)[0].str.replace('u', '')
treatments_clean['dose_end'] = treatments_clean['dose'].str.split(pat = '-', expand = True)[1].str.replace('u', '')
treatments_clean.drop(columns = 'dose', inplace = True)

##### Test

In [44]:
treatments_clean.shape

(350, 8)

In [45]:
treatments_clean

given_name        surname  hba1c_start  hba1c_end  hba1c_change  \
0        veronika       jindrová         7.63       7.20          0.43   
3            skye     gormanston         7.97       7.62          0.35   
6          sophia         haugen         7.65       7.27          0.38   
7           eddie         archer         7.89       7.55          0.34   
9            asia        woźniak         7.76       7.37          0.39   
10         joseph            day         7.70       7.19          0.51   
12        roxanne      andreyeva         9.54       9.14          0.40   
14         simone   baumgaertner         7.74       7.30          0.44   
15           enco         žibrik         7.78       7.34          0.44   
16        camilla       zaitseva         7.53       7.13          0.40   
21          tekla        walczak         7.61       7.29          0.32   
22    brancaleone          russo         8.61       8.18          0.43   
24           isac           berg         9.68       9.29          0.39   
28        clinton         miller         7.79       7.40          0.39   
29         eugene        mironov         7.81       7.48          0.33   
30    szilveszter          totth         7.70       7.38          0.32   
34      alexander      mathiesen         7.96       7.55          0.41   
36            chỉ            lâm         7.68       7.24          0.44   
37      władysław      wieczorek         7.92       7.47          0.45   
38       kristján        ingason         7.92       7.57          0.35   
42         marija       grubišić         7.53       7.15          0.38   
44          sauli     koivuniemi         7.67       7.37          0.30   
46        mariana          souza         7.86       7.51          0.35   
51     kristoffer      martinsen         9.18       8.64          0.54   
52             mỹ          quynh         7.61       7.16          0.45   
54           oles        zhdanov         7.52       7.11          0.41   
55        triana.       terrazas         7.71       7.34          0.37   
58         gabryŝ    tomaszewski         7.87       7.47          0.40   
59      leixandre         alanis         7.74       7.32          0.42   
60   onyekachukwu         obinna         7.58       7.12          0.46   
..            ...            ...          ...        ...           ...   
638          jane        citizen         7.98       7.60          0.38   
639        angela     lavrentyev         7.61       7.14          0.47   
640        edelma    villalpando         7.99       7.56          0.43   
642        annika          vaara         7.73       7.34          0.39   
644         chiho           higa         7.71       7.30          0.41   
652     beatrycze        woźniak         7.54       7.17          0.37   
653      miłosław     wiśniewski         7.51       7.08          0.43   
656       firenze          fodor         7.89       7.55          0.34   
657           zoe        wellish         7.71       7.30          0.41   
658           una  traustadóttir         8.00       7.50          0.50   
659         luboš          pecha         7.79       7.45          0.34   
663         meaza         brhane         7.70       7.36          0.34   
664         adlan       shishani         7.84       7.37          0.47   
667         sofia      hermansen         8.90       8.57          0.33   
668         guðni      heimisson         7.64       7.24          0.40   
670       eufemio        rosario         7.54       7.26          0.28   
671      dalmacia         madrid         7.67       7.21          0.46   
672         daimy          tromp         9.41       8.94          0.47   
674        jeremy        montagu         7.68       7.36          0.32   
678       nebechi      ekechukwu         7.78       7.39          0.39   
679       satsita      batukayev         7.63       7.25          0.38   
680       timothy         cotton         7.92       7.52          0.40   
681       b

In [45]:
treatments_clean.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 698
Data columns (total 6 columns):
given_name    350 non-null object
surname       350 non-null object
treatment     350 non-null object
dose          350 non-null object
dose_start    350 non-null object
dose_end      350 non-null object
dtypes: object(6)
memory usage: 19.1+ KB


#### Adverse reaction should be part of the `treatments` table

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution.*

##### Code

In [58]:
treatments_clean = pd.merge(treatments_clean, adverse_reactions, on = ['given_name', 'surname'], how = 'left')

##### Test

In [47]:
treatments_clean.shape

(350, 9)

In [48]:
treatments_clean.sample(20)

given_name        surname  hba1c_start  hba1c_end  hba1c_change  \
168            robert         maslov         7.65       7.35          0.30   
298           sargent        flamand         7.92       7.49          0.43   
10              tekla        walczak         7.61       7.29          0.32   
42             finley       chandler         7.65       7.26          0.39   
39               maya       isaksson         7.66       7.17          0.49   
235               mùi          lương         7.74       7.32          0.42   
83            benigna        ferrari         7.78       7.36          0.42   
122           johanna         dreher         7.77       7.34          0.43   
165             genet       fesahaye         7.73       7.36          0.37   
238              sabr          amari         7.94       7.57          0.37   
31            coralie        allaire         7.63       7.30          0.33   
349             berta     napolitani         7.68       7.21          0.47   
37              avdei       tikhonov         7.77       7.27          0.50   
74             harley         tucker         7.97       7.64          0.33   
55   zikoranaudodimma       chinedum         8.19       7.83          0.36   
229            markus        solberg         7.95       7.57          0.38   
2              sophia         haugen         7.65       7.27          0.38   
256        saidmohmad     rushisvili         7.90       7.48          0.42   
277        willem-jan  van der lubbe         7.57       7.20          0.37   
94               chân            bùi         7.53       7.18          0.35   

    treatment dose_start dose_end           adverse_reaction  
168   auralin        55        68                        NaN  
298   novodra        47        39                        NaN  
10    auralin        29        39                        NaN  
42    auralin        31        45                   headache  
39    auralin        33        41                        NaN  
235   novodra        52        50                        NaN  
83    auralin        41        52                        NaN  
122   auralin        51        59                        NaN  
165   auralin        25        35                        NaN  
238   novodra        32        27                        NaN  
31    auralin        39        46                        NaN  
349   novodra        42        44  injection site discomfort  
37    auralin        27        36                        NaN  
74    auralin        47        57                        NaN  
55    auralin        41        51                        NaN  
229   novodra        35        30                        NaN  
2     auralin        37        42                        NaN  
256   novodra        49        46                        NaN  
277   novodra        40        40                        NaN  
94    auralin        31        42                        NaN

In [49]:
treatments_clean[treatments_clean['given_name'] =='albinca']

given_name  surname  hba1c_start  hba1c_end  hba1c_change treatment  \
202    albinca  komavec         7.89       7.46          0.43   novodra   

    dose_start dose_end adverse_reaction  
202        41        39     hypoglycemia

In [50]:
adverse_reactions[adverse_reactions['given_name'] =='albinca']

given_name  surname adverse_reaction
7    albinca  komavec     hypoglycemia

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for one function used in the solution and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html) for another function used in the solution.*

##### Code

In [59]:
treatments_clean['given_name'] = treatments_clean['given_name'].str.title()
treatments_clean['surname'] = treatments_clean['surname'].str.title()

treatments_clean = pd.merge(left = treatments_clean,  right = patients[['given_name', 'surname', 'patient_id']], 
                            on = ['given_name', 'surname'], how='left')

treatments_clean.drop(columns = ['given_name', 'surname'], inplace = True)

cols = ['patient_id', 'treatment', 'dose_start','dose_end', 'hba1c_start', 'hba1c_end', 'hba1c_change', 'adverse_reaction', ]
treatments_clean =  treatments_clean[cols]

In [52]:
treatments_clean.columns

Index(['patient_id', 'treatment', 'dose_start', 'dose_end', 'hba1c_start',
       'hba1c_end', 'hba1c_change', 'adverse_reaction'],
      dtype='object')

##### Test

In [60]:
treatments_clean.shape

(350, 8)

In [61]:
treatments_clean.sample(10)

patient_id treatment dose_start dose_end  hba1c_start  hba1c_end  \
264       408.0   novodra        34        34         7.79       7.43   
228       414.0   novodra        37        36         7.81       7.42   
34        449.0   auralin        25        32         9.12       8.73   
45         65.0   auralin        39        45         7.73       7.42   
158       159.0   auralin        28        33         7.64       7.29   
60        195.0   auralin        30        38         7.60       7.16   
175        94.0   novodra        40        45         7.56       7.09   
63         61.0   auralin        29        45         7.90       7.46   
92         99.0   auralin        53        57         7.57       7.19   
232       413.0   novodra        31        30         7.51       7.17   

     hba1c_change adverse_reaction  
264          0.36         headache  
228          0.39              NaN  
34           0.39              NaN  
45           0.31              NaN  
158          0.35              NaN  
60           0.44              NaN  
175          0.47     hypoglycemia  
63           0.44              NaN  
92           0.38              NaN  
232          0.34              NaN

### Save the dataframes

In [64]:
patients_clean.to_csv('patients_clean.csv', index = False)
treatments_clean.to_csv('treatments_clean.csv', index = False)

In [4]:
patients_clean = pd.read_csv('patients_clean.csv')
treatments_clean = pd.read_csv('treatments_clean.csv')

### Quality

<font color='red'>Complete the remaining "Quality" **Define, Code, and Test** sequences after watching the *"Cleaning for Quality"* video.</font>

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
*Your definition here. Hint: see the "Data Cleaning Process" page.*

##### Code

In [11]:
patients.zip_code = patients.zip_code.astype('str').str[:-2].str.pad(width = 5, side = 'left', fillchar = '0')

##### Test

In [ ]:
# Your testing code here

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
*Your definition here.*

##### Code

In [16]:
patients_clean[patients_clean['surname'] == 'Neudorf'].height = 72

/opt/conda/lib/python3.6/site-packages/pandas/core/generic.py:4405: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [26]:
patients_clean['surname'] == 'Neudorf'

0      False
1      False
2      False
3      False
4       True
5      False
6      False
7      False
8      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18     False
19     False
20     False
21     False
22     False
23     False
24     False
25     False
26     False
27     False
28     False
29     False
       ...  
473    False
474    False
475    False
476    False
477    False
478    False
479    False
480    False
481    False
482    False
483    False
484    False
485    False
486    False
487    False
488    False
489    False
490    False
491    False
492    False
493    False
494    False
495    False
496    False
497    False
498    False
499    False
500    False
501    False
502    False
Name: surname, Length: 503, dtype: bool

In [28]:
patients_clean.loc[4,'height'] = 72

##### Test

In [ ]:
# Your testing code here

#### Full state names sometimes, abbreviations other times

##### Define
*Your definition here. Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.*

##### Code

In [30]:
patients['state'].value_counts()

California    36
TX            32
New York      25
CA            24
MA            22
NY            22
PA            18
GA            15
Illinois      14
OH            14
Florida       13
LA            13
MI            13
OK            13
NJ            12
VA            11
WI            10
IL            10
MS            10
TN             9
IN             9
MN             9
FL             9
AL             9
KY             8
NC             8
WA             8
MO             7
KS             6
NV             6
ID             6
CT             5
IA             5
SC             5
AR             4
ME             4
AZ             4
Nebraska       4
RI             4
CO             4
ND             4
OR             3
WV             3
DE             3
MD             3
SD             3
MT             2
DC             2
NE             2
VT             2
WY             1
NM             1
AK             1
NH             1
Name: state, dtype: int64

In [34]:
def state_abrv(state):
    if state == 'California':
        return 'CA'
    elif state == 'New York':
        return 'NY'
    elif state == 'Illinois':
        return 'IL'
    elif state == 'Florida':
        return 'FL'
    elif state == 'Nebraska':
        return 'NE'
    else :
        return state

In [36]:
patients_clean['state'] = patients_clean['state'].apply(state_abrv)

##### Test

In [ ]:
# Your testing code here

#### Dsvid Gustafsson

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
*Your definition here. Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) for one method used in solution, [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) for one function used in the solution, and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) for another method used in the solution.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Multiple phone number formats

##### Define
*Your definition here. Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Default John Doe data

##### Define
*Your definition here. Recall that it is assumed that the data that this John Doe data displaced is not recoverable.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here

#### kgs instead of lbs for Zaitseva weight

##### Define
*Your definition here.*

##### Code

In [ ]:
# Your cleaning code here

##### Test

In [ ]:
# Your testing code here